jetty-io/src/main/java/org/eclipse/jetty/io/ManagedSelector.java

```diff
@@ -115,28 +115,67 @@ protected void doStop() throws Exception
     {
         if (LOG.isDebugEnabled())
             LOG.debug("Stopping {}", this);
-        CloseEndPoints close_endps = new CloseEndPoints();
-        submit(close_endps);
-        close_endps.await(getStopTimeout());
-        CloseSelector close_selector = new CloseSelector();
-        submit(close_selector);
-        close_selector.await(getStopTimeout());
-
-        super.doStop();
         
+        try
+        {
+            CloseAction close = new CloseAction();
+            submit(close);
+            if (getStopTimeout()>0)
+                close.await(getStopTimeout());
+        }
+        finally
+        {
+            try
+            {
+                super.doStop();
+            }
+            finally
+            {
+                try (Locker.Lock lock = _locker.lock())
+                {
+                    if (_selector!=null && _selector.isOpen())
+                        _selector.wakeup();
+
+                    _actions.stream().filter(CloseAction.class::isInstance).forEach(a->a.run());
+                    _actions.clear();
+                }
+            }
+        }
         if (LOG.isDebugEnabled())
             LOG.debug("Stopped {}", this);
     }
 
-    public void submit(Runnable change)
+    @Deprecated
+    public void submit(Runnable action)
```

Comment:
IMHO this method should be package-private or protected, not public.

I would keep it as `submit(Runnable) ` because actions are treated as Runnables anyway (where we poll them), so while I understand the meaning of "you can only submit Actions, not Runnables" in changing the signature, I think it just buys us very little and clutters the code to have 2 copy of the same method.

Reply:
I think this is needed to avoid the little mess we made by submitting Blocking runnables.   However I agree that the new submit(Action) can be protected.